src/librustdoc/clean/types.rs

```diff
@@ -743,67 +741,111 @@ impl Item {
-                out.push(&align_s);
-            }
-            let int_s;
-            if let Some(int) = repr.int {
-                int_s = match int {
-                    IntegerType::Pointer(is_signed) => {
-                        format!("{}size", if is_signed { 'i' } else { 'u' })
+        if !keep_as_is && let Some(repr) = self.repr(tcx, cache) {
+            attrs.push(repr);
+        }
+
+        attrs
+    }
+
+    /// Compute the *public* `#[repr]` of this item.
+    ///
+    /// Read more about it here:
+    /// https://doc.rust-lang.org/nightly/rustdoc/advanced-features.html#repr-documenting-the-representation-of-a-type
+    fn repr<'tcx>(&self, tcx: TyCtxt<'tcx>, cache: &Cache) -> Option<String> {
+        let def_id = self.def_id()?;
+        let (ItemType::Struct | ItemType::Enum | ItemType::Union) = self.type_() else {
+            return None;
+        };
+
+        let adt = tcx.adt_def(def_id);
+        let repr = adt.repr();
+
+        let is_visible = |def_id| cache.document_hidden || !tcx.is_doc_hidden(def_id);
+        let is_field_public =
+            |field: &'tcx ty::FieldDef| field.vis.is_public() && is_visible(field.did);
+
+        if repr.transparent() {
+            // `repr(transparent)` is public iff the non-1-ZST field is public or
+            // at least one field is public in case all fields are 1-ZST fields.
+            let is_public = cache.document_private
+                || adt.variants().iter().all(|variant| {
+                    if !is_visible(variant.def_id) {
+                        return false;
                     }
-                    IntegerType::Fixed(size, is_signed) => {
-                        format!("{}{}", if is_signed { 'i' } else { 'u' }, size.size().bytes() * 8)
+
+                    let field = variant.fields.iter().find(|field| {
+                        let args = ty::GenericArgs::identity_for_item(tcx, field.did);
+                        let ty = field.ty(tcx, args);
+                        tcx.layout_of(tcx.param_env(field.did).and(ty))
+                            .is_ok_and(|layout| !layout.is_1zst())
+                    });
+
+                    if let Some(field) = field {
+                        return is_field_public(field);
                     }
-                };
-                out.push(&int_s);
-            }
-            if !out.is_empty() {
-                attrs.push(format!("#[repr({})]", out.join(", ")));
-            }
+
+                    adt.variants().iter().all(|variant| {
+                        variant.fields.is_empty() || variant.fields.iter().any(is_field_public)
+                    })
+                });
+
+            // Since `repr(transparent)` can't have any other reprs or
+            // repr modifiers beside it, we can safely return early here.
+            return is_public.then(|| "#[repr(transparent)]".into());
         }
-        attrs
+
+        let is_public = cache.document_private
```

Comment:
I should return early here if it's the default repr / `repr(Rust)` for perf reasons.